pkg/server/singleprocess/service_pipeline.go

```diff
@@ -118,26 +117,163 @@ func (s *Service) RunPipeline(
 	}); err != nil {
 		return nil, err
 	}
-	run, err := s.state(ctx).PipelineRunGetLatest(pipeline.Id)
+
+	pipelineRun, err := s.state(ctx).PipelineRunGetLatest(pipeline.Id)
+	if err != nil {
+		return nil, err
+	}
+
+	// Build out all of the queued job requests for running this pipelines steps
+	stepJobs, pipelineRun, stepIds, err := s.buildStepJobs(ctx, log, req,
+		make(map[string]interface{}), nodeIdMap, make(map[string][]string), pipeline, pipelineRun)
 	if err != nil {
 		return nil, err
 	}
 
-	// Generate the jobs for each of the steps
+	// Get the ordered jobs.
+	var jobIds []string
+	jobMap := map[string]*pb.Ref_PipelineStep{}
+	for _, v := range stepGraph.KahnSort() {
+		// Look up step name and ref by the assigned node ID from graph generation
+		nodeId := v.(string)
+		stepRef, ok := nodeIdMap[nodeId]
+		if !ok {
+			return nil, status.Errorf(codes.Internal,
+				"could not get pipeline step ref for node id %q", nodeId)
+		} else if stepRef == nil {
+			return nil, status.Errorf(codes.Internal,
+				"node id %q returned a nil pipeline step ref", nodeId)
+		}
+
+		// get the generated queued job request
+		jobId, ok := stepIds[nodeId]
+		if !ok {
+			// NOTE(briancain):
+			// This could be better. It's basically here because we want to keep track
+			// of an embedded pipeline step ref within the step graph, but it doesn't have
+			// a "job id" because the root of the embedded pipeline is the actual ID where
+			// this is simply a reference.
+			// We don't want to add it as a job id because it doesn't actually create a job.
+			// jobId = "embedded-pipeline-ref-" + nodeId
+			continue
+		}
+
+		jobIds = append(jobIds, jobId)
+		stepRef.RunSequence = pipelineRun.Sequence
+		jobMap[jobId] = stepRef
+	}
+
+	pipelineRun.State = pb.PipelineRun_STARTING
+	if err = s.state(ctx).PipelineRunPut(pipelineRun); err != nil {
+		return nil, err
+	}
+
+	// Queue all the jobs atomically
+	if _, err := s.queueJobMulti(ctx, stepJobs); err != nil {
+		return nil, err
+	}
+
+	return &pb.RunPipelineResponse{
+		JobId:     jobIds[0],
+		AllJobIds: jobIds,
+		JobMap:    jobMap,
+	}, nil
+}
+
+func (s *Service) buildStepJobs(
+	ctx context.Context,
+	log hclog.Logger,
+	req *pb.RunPipelineRequest,
+	visitedPipelines map[string]interface{},
+	nodeIdMap map[string]*pb.Ref_PipelineStep,
```

Comment:
Idea, what if we switched this to:

`map[*pb.Ref_PipelineStep]string`? Then instead of adding another nested loop in `stepToNodeId` to loop through the whole map for a value match, we can do an O(1) look up and grab the `nodeId`?

Reply:
I believe I initially attempted this structure, where a protobuf was the `key` for the map, but you can't safely compare protobuf messages as values I _don't think_. I know Golang does allow for regular Go structs to be used as keys for a map key, but as far as a Protobuf message goes it's recommended to use [proto.Equal](https://developers.google.com/protocol-buffers/docs/reference/go/faq#deepequal) for comparison and I'm not sure how that behaves with maps, thus the structure here.

```golang
stepRef := &pb.Ref_PipelineStep {
  Name: "test",
}
mymap := map[*pb.Ref_PipelineStep]string{stepRef: "123abc"}

// Not sure about this one:
fmt.Println(mymap[stepRef])
```

If this is safe and reliable to do though it would make this code simpler for sure and would solve the issue you mentioned later in this PR about the node ID lookup loop iterating over this map.

Reply:
Ok, @evanphx and I figured out a decent solution to this! Solved via [d524338](https://github.com/hashicorp/waypoint/pull/3584/commits/d5243388e6c1979bd0d49ca1ee71bea3b52368a7)